subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy

```diff
@@ -158,4 +161,60 @@ class WrapperGenerationIntegrationTest extends AbstractIntegrationSpec {
             getValue(Attributes.Name.IMPLEMENTATION_TITLE) == 'Gradle Wrapper'
         }
     }
+
+    @Rule
+    HttpServer httpServer = new HttpServer()
+
+    def "wrapper task fails if distribution url from command-line is invalid"() {
+        given:
+        def path = "/distributions/8.0-RC-5"
+        httpServer.start()
+        httpServer.expectHeadMissing(path)
+        def url = "${httpServer.uri}" + path
+
+        when:
+        run "wrapper", "--gradle-distribution-url", url
+
+        then:
+        Throwable throwable = thrown()
+        def exception = throwable.cause.cause.cause
+        assert exception.class == UncheckedIOException.class
+        assert exception.message == "Test of distribution url failed. Please check the values set with --gradle-distribution-url and --gradle-version."
+
+        def wrappedException = exception.cause
+        assert wrappedException.class == RuntimeException.class
+        assert wrappedException.message == "HEAD request to " + url + " failed: response code (404)"
+        file("gradle/wrapper/gradle-wrapper.properties").assertDoesNotExist()
+    }
+
+    def "wrapper task succeeds if distribution url from command-line is valid"() {
+        given:
+        def path = "/distributions/8.0-rc-5"
+        def file = file(path) << "some content"
+        httpServer.start()
+        httpServer.expectHead(path, file)
+        def url = "${httpServer.uri}" + path
+
+        when:
+        run "wrapper", "--gradle-distribution-url", url
+
+        then:
+        succeeds()
+    }
+
+    def "wrapper task with distribution url from command-line respects --offline"() {
+        when:
+        run("wrapper", "--gradle-distribution-url", "https://not-a-valid-url/distributions/8.0-RC-5", "--offline")
+
+        then:
+        succeeds()
+    }
+
+    def "wrapper task with gradle version from command-line respects --offline mode"() {
+        when:
+        run "wrapper", "--gradle-version", "8.0-RC-5", "--offline"
+
+        then:
+        succeeds()
```

Comment:
:lipstick: it would be nice if these tests would start the http server and assert no requests are being made at all